runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -1,14 +1,269 @@
+                Preconditions.checkState(e.getCause() instanceof TrimmedException);
+
+                // This means that there have been updates to the Replication Status table after the above timestamp,
+                // and this snapshot has been evicted from the cache of the JVM.  We cannot tell the status of snapshot
+                // sync at this timestamp.  Return true so that caller retries the check with a later timestamp.
+                return true;
+            }
+        }
+
+        /**
+         * Wait for the snapshot sync to complete.  The completion is signalled when SnapshotSyncCompletionListener
+         * invokes notifyAll().
+         * @param timestamp
+         */
+        private void waitSnapshotSyncCompletion(CorfuStoreMetadata.Timestamp timestamp) {
+            SnapshotSyncCompletionListener snapshotSyncCompletionListener = new SnapshotSyncCompletionListener(timestamp);
+            corfuStore.subscribeListener(snapshotSyncCompletionListener, CORFU_SYSTEM_NAMESPACE, LR_STATUS_STREAM_TAG,
+                    Arrays.asList(REPLICATION_STATUS_TABLE), timestamp);
+
+            synchronized (timestamp) {
+                while (!snapshotSyncCompletionListener.isSnapshotSyncComplete()) {
+                    try {
+                        timestamp.wait();
+                    } catch (Exception e) {
+                        log.error("Exception while waiting for snapshot sync to complete", e);
+                        corfuStore.unsubscribeListener(snapshotSyncCompletionListener);
+                        throw new RuntimeException(e);
+                    }
+                }
+                corfuStore.unsubscribeListener(snapshotSyncCompletionListener);
+            }
+        }
+    }
+
+    private class SnapshotSyncCompletionListener implements StreamListener {
+
+        private CorfuStoreMetadata.Timestamp subscriptionTimestamp;
+
+        @Getter
+        private boolean snapshotSyncComplete;
+
+        SnapshotSyncCompletionListener(CorfuStoreMetadata.Timestamp subscriptionTimestamp) {
+            this.subscriptionTimestamp = subscriptionTimestamp;
+            this.snapshotSyncComplete = false;
+        }
+
+        @Override
+        public void onNext(CorfuStreamEntries results) {
+            Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+            for (TableSchema tableSchema : entries.keySet()) {
+                if (tableSchema.getTableName().equals(REPLICATION_STATUS_TABLE)) {
+                    for (CorfuStreamEntry entry : entries.get(tableSchema)) {
+                        ReplicationStatusVal status = (ReplicationStatusVal) entry.getPayload();
+                        if (status.getDataConsistent()) {
+                            // Snapshot Sync has ended.  Notify the waiting thread to continue
+                            synchronized (subscriptionTimestamp) {
```

Comment:
Why do we need to synchronize? The subscriptionTimestamp doesn't seem to get modified once passed to the listener.
Or did I miss something?

Reply:
good question.  wait() and notify() are invoked inside a synchronized block to make sure that the thread owns the object's monitor -
https://stackoverflow.com/questions/52764462/how-to-use-wait-and-notify-with-the-use-of-different-classes-in-java

Reply:
nit: how about using countDownLatch then?

Reply:
+1 on countDownLatch.. Also, should we call it with a timeout? Cause what happens if a full sync is stuck, never completed because LR is down? what should we do in those scenarios?

Reply:
changed to using a countdown latch.

Calling it with a timeout of 10 seconds.  If snapshot sync apply does not finish within this time, we retry

Reply:
changed this to 1 min timeout as discussed offline and made the timeout a runtime parameter.